File: BOJ/10799_쇠막대기.cpp
#include <iostream>
#include <stack>
using namespace std;

int main()
{
	int result = 0;
	bool flag = false;
	stack<char> st;
	string pipe;
	cin >> pipe;

	for (int i = 0; i < pipe.length(); i++)
	{
		if (pipe[i] == '(')
		{
			++result;
			st.push('(');
			flag = true;
		}
		else
		{
			if (flag)
			{
				st.pop();
				result = result - 1 + st.size();
			}
			else
			{
				st.pop();
			}
			flag = false;
		}
	}
	cout << result << '\n';

	return 0;
}

File: BOJ/10799_쇠막대기.py
import sys

pipe = sys.stdin.readline().rstrip()
result = 0
prev = False
st = []

for i in pipe:
    if i == '(':
        result += 1
        st.append(i)
        prev = True
    else:
        if prev is True:
            st.pop()
            result += (len(st)-1)
        else:
            st.pop()
        prev = False
print(result)